Copy Image to Clipboard

jcjgraf opened the pull request:
Adds a new command `copy-image` to copy the whole image pixmap to the system clipboard, allowing easy pasting to other applications. `yi` and `yI` are currently bound to this command (global clipboard and selection clipboard respectively). 

Todo:
- [x] Check if implementation fulfils requirements
- [ ] ~Flag to copy the original image in contrast to the (potentially compressed) QPixmap~
- [x] Unit tests?

Implements: #358

wedens commented:
I've tried copying using the new command and pasting to several apps (some of them are: discord, pinta, rambox, qutebrowser) and it works as expected.

karlch commented:
Thanks a lot for the PR and sorry for being inactive recently, let's just say life is currently a bit weird :laughing: 

The implementation LGTM, what would remain is a quick test and an update to the changelog, but I am happy to look into this after merging if you prefer. I should find quite some time to work on vimiv over the week-end :blush:

jcjgraf commented:
Thanks @wedens for your testing and feedback!

As pointed out by @pglira and briefly discussed in #358, copying does not work for some images and target applications. I have the feeling that this is related to limitations in image size of the target application and hence nothing which we can fix.

Currently, the QPixmap opened by vimiv, is copied and **not** the original images file. Therefore, metadata may be lost and the image quality may also deteriorate.

Therefore, I thought it would be a good addition to allow copying the original file through a flag for `copy-image`. I have come up with the following snippet:

```Python
url = QUrl.fromLocalFile(path)
mime_data = QMimeData()
mime_data.setUrls([url])
clipboard.setMimeData(mime_data, mode=mode)
```

In (scientifically meaningless) tests (due to the lack of applications which allow image pasting) this method only worked for Libre Office Write and not for Element (Riot) :thinking:  Therefore, I am unsure how widely supported this method is.

Unless someone has a solution (or can confirm that this method actually works well in other applications) I would suggest to stick to the current implementation and potentially open an issue to remind us to add this functionality some point in the future.

What do you reckon @karlch @pglira?

karlch commented:
I think I prefer the current implementation. If one really wants all the information of the current image in its original state, one can always copy the abspath and deal with it in the other application accordingly. I would assume this command would be used more for simple inserting of the image and it would probably be resized (libreoffice) or compressed (web applications) afterwards in any case. If you assume the actual image data size is an issue, could it make sense to add flags for `--width` and `--height` to rescale the `QImage` before copying? This could also come in handy to not have a huge image inserted into libreoffice. What do you think?

jcjgraf commented:
Thanks for the feedback and this great idea!

I have added the flag `width` and `height` to scale the image to the specified with **OR** height respectively. It is not possible to scale it to a specified wight and hight. Should that be possible?
I have also added the `size` flag which scales the longer side of the image to the specified value.

There is a hardcoded precedence if the user specifies multiple flags. Should an error or something be thrown in such a case or is it alright how it is implemented?

Let me know if you have a neater way to implement this! Also tell if you prefer the code to be more compact (i.e. less empty lines :upside_down_face: )

jcjgraf commented:
I have tried to put some meaningful end2end tests for this new command together. For some reason the CI is not happy with them even though they work for me locally :confused: 

I am not very familiar with writing test, so feedback is very welcome :blush:

One question I have is what images can I open with `I open the image <image>`? I have just selected `.hidden.jpg` since this is the only one you have used in this context.

karlch commented on tests/end2end/features/misc/clipboard.feature:
```diff
@@ -26,3 +26,28 @@ Feature: Interaction with the system clipboard.
         Then the working directory should be child_01
+
+    Scenario: Copy image to clipboard.
+        Given I open the image '.hidden.jpg'
```
This `Given` step allows defining the name of the image we open, so you can use any name you like. If you don't need a specific name, you can just use `Given I open any image` which will create `image.jpg`.

karlch commented on tests/end2end/features/misc/conftest.py:
```diff
@@ -23,3 +23,33 @@ def check_clipboard(clipboard, text):
+
+@bdd.then(bdd.parsers.parse("The clipboard should contain image {image}"))
+def check_clipboard_image(clipboard, image):
+    assert clipboard.pixmap(mode=QClipboard.Clipboard).toImage() == QImage(image)
```
I asssume this test fails for some reason under qt 5.9 and 5.10. I would be happy to loosen this to `should contain any image` where we just ensure that we successfully saved an image to the clipboard without comparing the content directly.

jcjgraf replied:
This seems to work. Changed that in 0fb4369

karlch commented on tests/end2end/features/misc/conftest.py:
```diff
@@ -23,3 +23,33 @@ def check_clipboard(clipboard, text):
+
+@bdd.then(bdd.parsers.parse("The primary selection should contain image {image}"))
+def check_primary_image(clipboard, image):
+    assert clipboard.pixmap(mode=QClipboard.Selection).toImage() == QImage(image)
```
Same as above :blush:

jcjgraf replied:
Changed in 0fb4369

karlch commented on vimiv/api/_modules.py:
```diff
@@ -89,6 +89,54 @@ def copy_name(abspath: bool = False, primary: bool = False) -> None:
+@api.keybindings.register("yI", "copy-image --primary")
+@api.commands.register()
+def copy_image(
+    primary: bool = False, width: int = -1, height: int = -1, size: int = -1
```
I would personally prefer to use `None` instead of `-1` as it fits better to the `Optional` character of this flag.

Another random idea I had is to support `count` here which would set the size using `count` so e.g. `800yi` would result in the same behaviour as `:copy-image --size 800` to allow setting the size without interacting with the command line.

jcjgraf replied:
Imrpoved the None in cdbacf4

Added the count support in d767e5b55d45f14ba22f5cde86ad00f337d3c4a4. Have I added the documentation for the count correctly to the docstring?

karlch replied:
Looks good to me except for possibly a missing whitespace after `--size`.

jcjgraf replied:
good catch, fixed in 421e0c65c372d0672455c22bcfd9a32d602e16e4

karlch reviewed (commented):
Thanks a lot, this looks great! I have added a few minor comments up for discussion and to clarify the CI :blush:

karlch commented on vimiv/api/_modules.py:
```diff
@@ -89,6 +89,62 @@ def copy_name(abspath: bool = False, primary: bool = False) -> None:
+
+        if pix_size.height() >= pix_size.width():
+            _logger.debug(f"Copy image with size {size} restricting height")
+            pixmap = pixmap.scaledToHeight(size)
```
Feel free to add `  # type: ignore[arg-type]` to the two lines failing mypy. It probably cannot deal with this slightly non-trivial check to know that size is now certainly an `int` and no longer possibly `None`.

With this and splitting the syntax section into two lines
```
    **syntax:** ``:copy-image [--primary] [--width=WIDTH] [--height=HEIGHT]
    [--size=SIZE]``
```
the nasty CI should finally be happy (except for the annoying timeout issue) :laughing:

jcjgraf replied:
All green, thanks for the hints!
